### PR TITLE
Only memorize recipes when skill greatly exceeds requirements.

### PR DESCRIPTION
**Pull request by kevingranade**
As per discussion in #7866, the primary way to memorize recipes should be to make them, with memorization from books only serving to backfill recipes you far exceed the requirements for.

### CODE REVIEW

File: src/player.cpp [hunk 1]
```diff
@@ -11258,7 +11254,7 @@ bool player::can_study_recipe(const itype &book) const
         auto const r = elem.recipe;
         if( !knows_recipe( r ) &&
             ( !r->skill_used ||
-              get_skill_level( r->skill_used ) >= elem.skill_level ) ) {
+              ( get_skill_level( r->skill_used ) - elem.skill_level ) >= 4 ) ) {
```
Comment (mugling) [inline src/player.cpp]: This results in you needing to exceed the requirements by 4 levels to memorize a recipe? Could be factored out as a constant and possibly reduced to 3?

**Comment by Coolthulhu** [comment]
How about having 2 separate, non-hardcoded levels of book recipes?

Book recipes are already arrays, like this: `[ "book_id", level ]`. They could be easily extended into `[ "book_id", use_level, memorize_level ]`, with the latter defaulting to `use_level + x`.

**Comment by mugling** [comment]
> Book recipes are already arrays, like this: [ "book_id", level ]. They could be easily extended into [ "book_id", use_level, memorize_level ], with the latter defaulting to use_level + x.

That could work well

**Comment by Rod-Serling** [comment]
Having a hard memorize level feels off to me. If you can make it, and you make it enough, you should be able to memorize it even if you don't necessarily have the level. (I made 20 of them, then four days later after skill rust reset my leveling, I made 20 more, but I still don't have it memorized)

I'd rather see an RNG check. At use_level + 4 it could be like a 90% chance or something. At use_level it might only be a 10% chance.

File: src/player.cpp (hunk 1 above)
Comment (kevingranade) [inline src/player.cpp]: > factored out as a constant

Sure, though now that I think of it I think I'll factor out a method instead.
> and possibly reduced to 3?

Any particular rationale? I'm not saying 4 is perfect, it's just the middle of the previous random range, I'm just wondering if you have a reason.

**Comment by kevingranade** [comment]
> How about having 2 separate, non-hardcoded levels of book recipes?

I guess I can add that, but is anyone going to be interested in doing anything with it other than leaving it at the default?  It seems like a hell of a lot of work for negligible benefit.
> If you can make it, and you make it enough, you should be able to memorize it

You do, the primary way to memorize recipes is by performing them, this just relegates the process of memorizing recipes from reading books to backfilling your known recipes quickly if you're very overleveled for the recipe.  The only reason I'm leaving memorization from books in at all is it's something you should in principle be able to do, it serves very nearly no gameplay purpose.

**Comment by Isaacssv552** [comment]
Why doesn't this use the the skill threshold as per the discussion? It would be both odd and annoying not to be able to book-memorize a recipe that is so easy you get no xp.

**Comment by kevingranade** [comment]
Bumping as I still consider this ready to go, no one has expressed any support in the feature Coolthulu requested, including Coolthulu.

> Why doesn't this use the the skill threshold as per the discussion?

Because on examination that means you go from "just learned to do it" to "can memorize the recipe by reading it once" over the course of a very short practice session, especially for low-level recipes.
The effect of this would be that as soon as you reach level 2 in a skill, you can immediately memorize every lvl 1 recipe, and this continues for lvls 2 and 3, and only widens to a two-level gap at level 4.

The reason this gap was small at first and then grew for skill cap is that we could afford to make it small at first to have a high impact (you can only grind on a particular recipe for a level or so), but at higher levels the gap needed to widen because the number of recipes available drops, and players need enough selection to ensure they can make progress.

For memorization, we have no such requirements, it simply needs to have a gap that makes sense, and there are no balance issues as reading is the secondary way to memorize recipes.

File: src/player.cpp (hunk 1 above)
Comment (mugling) [inline src/player.cpp]: > Sure, though now that I think of it I think I'll factor out a method instead.

Either would work and provide more documentation than a literal constant

**Comment by Coolthulhu** [comment]
4 level gap is wide enough to make book memorization not matter, but not wide enough to make it not happen. It makes it look like a feature that is kept for the sake of not throwing away features.
It is neither realistic (being able to craft custom kevlar armor but not book-memorize cargo pants, crafting sci-fi UPS but not memorizing signal receivers) nor does it leave any room for any gameplay impact of book memorization (at skill level 5 you most likely have a car for the books).

Honestly I'd prefer if it went away totally and we'd have one section of code less to maintain.

**Comment by kevingranade** [comment]
I'll rebase the original pr then, that was the first plan :)

To be clear though, is it just the gap or the whole idea? There was a
suggestion to drop it to 3, and I wasn't opposed to adjusting it, but
wanted a rationale.

**Comment by mugling** [comment]
> To be clear though, is it just the gap or the whole idea?

The idea is reasonable, the gap is the issue. The selection of 3 was somewhat arbitrary and @Coolthulhu has done a better job of outlining the rationale.

It might be simpler to just drop memorization from reading?

**Comment by Coolthulhu** [comment]
Obsoleted by #17161

**Pull request closed without merging**